python/ml4ir/base/model/architectures/auto_dag_network.py

```diff
@@ -0,0 +1,362 @@
+        if module == "builtins":
+            return cls.__qualname__  # avoid outputs like 'builtins.str'
+        return f"{module}.{cls.__qualname__}"
+
+    def get_sub_classes(cls):
+        """DFS traversal to get all subclasses of `cls` class"""
+        for subcls in cls.__subclasses__():
+            if subcls not in subclasses:
+                # Handles duplicates
+                subclasses[full_class_name(subcls)] = subcls
+                get_sub_classes(subcls)
+
+    subclasses = {}
+    get_sub_classes(layers.Layer)
+    return subclasses
+
+
+class CycleFoundException(Exception):
+    """Thrown when a cycle is identified in a DAG"""
+    pass
+
+
+class LayerNode:
+    """Defines a node in the `LayerGraph`"""
+
+    def __init__(
+            self,
+            name: str,
+            layer: layers.Layer = None,
+            dependent_children: List[LayerNode] = None,
+            inputs: List[str] = None,
+            inputs_as_list: bool = False
+    ):
+        """
+        Constructor to create a node
+
+        Parameters
+        ----------
+        name: str
+            Name of the node
+        layer: instance of type tf.keras.layers.Layer
+            Node operation, i.e. Layer for forward prop
+        dependent_children: list of LayerNode
+            All downstream nodes which depend on this node's output
+        inputs: list of str
+            All inputs to the `layer` to be passed during forward prop
+        inputs_as_list: bool
+            Boolean to indicate if the set of input tensors to the layer should be converted to list
+        """
+        self.name = name
+        self.layer = layer
+        self.dependent_children = dependent_children if dependent_children is not None else []
+        self.inputs = inputs
+        self.is_input_node = True if inputs is None else False
+        self.inputs_as_list = inputs_as_list
+
+    def __str__(self):
+        """Node is identified by its name"""
+        return self.name
+
+    def __repr__(self):
+        """Get the obj representation"""
+        return self.name
+
+
+class LayerGraph:
+    """Defines a DAG of `LayerNode`"""
+
+    def __init__(
+            self,
+            layer_ops: Dict[str, Union[str, bool, layers.Layer]],
+            inputs: List[str]
+    ):
+        """
+        Constructor to create a Graph. It creates a dependency graph and identifies the output node.
+
+        Parameters
+        ----------
+        layer_ops: dict
+            Dictionary which contains information for all the layer nodes in the graph. Expected schema:
+            {
+                "NodeName":  # Name of the layer node
+                {
+                    "inputs": ["InputNodeName1", "InputNodeName2"],  # Name of the input layer nodes for this layer
+                    "op": keras.layers.Layer.ChildClass(),  # Layer instance which defines the operation of the node
+                    "aslist": True  # If inputs to the "op" should be a list of tensors
```

Comment:
@arvindsrikantan Update docs with `args`

Reply:
There are no args here

Reply:
Yeah, I realized that later.